### DISCUSSION TITLE
Some more ideas and comments

### DISCUSSION THREAD

**Discussion by rainymatch** [ArturoDent/jump-and-select#10] [Ideas]
1) It would be great to have the option to set different behaviour for jumping and selection: 
```
  "jump-and-select.putCursorForward"      :      "beforeCharacter",
  "jump-and-select.putCursorBackward"     :      "afterCharacter",
```
(For me, the expected/intuitive behaviour is to jump before the character, while selection should be inclusive. Most importantly, I don't feel like the behaviour should necessary be the same for both actions.)

2) In scenarios where I want to jump to the next or previous character, while being at the moment before/after an instance of such character, I think the very next character should be ignored. (It does not make much sense to jump just by one character, instead of using arrows.)
Imagine here having the cursor in between the last pair of parents `all(|)` - if I want to jump to the `|(Topic)` parenthesis, I need to move by one character to the left and then use the jump functionality. At the same time, it seems obvious that I don't want to jump to `all|()`... 
```
topics = db.query(Topic).all()
```

3) In MultiMode, it would be super neat to have the option to go either backwards or forward - not just the one direction. I undrestand this might be not really easy to implement, but still. May be, the solution would be to have a keybinding to exit the multimode and reenter it in the opposite direction...

**Comment by ArturoDent** [comment]
To clarify for (3) you would like to **reverse** directions but stay in (or re-enter) multiMode?  

You can do this keybinding

```json
{
  "key": "alt+r",
  "command": "jump-and-select.jumpBackwardMultiMode"
}
```
 which will revese the multiMode when going forward to backward but you would have to have one for `jumpBackwardMultiMode` as well.  So something like

`jump-and-select.reverseMultiMode` would work.

I wonder if it would be possible to set a character (or something like alt+shift) that would trigger the reverse - which would be very handy?  I'll investigate.

**Comment by ArturoDent** [comment]
(1)  You can do this now with these keybindings:

```jsonc
{
  "key": "alt+r",                        // whatever you want
  "command": "jump-and-select.jumpForward",
  "args": {
    "putCursorForward": "beforeCharacter"
  }
},

{
  "key": "shift+alt+r",                 // whatever you want
  "command": "jump-and-select.jumpForwardSelect",
  "args": {
    "putCursorForward": "afterCharacter"
  }
}
```  

You'll get a warning squiggly under `args` because there is no `text` option - in the JSON Schema I set `text` to `required` but I don't see a reason not to relax that.  I'll think about it but will probably make that change.  In the meantime the above keybindings work despite the squiggly.  

You would have to make similar keybindings for all the other commands though - backward and multiMode and select - so would 2 settings like 

```plaintext
jump: beforeCharacter
select: afterCharacter
```

be enough or would the `jumpBackward` and `jumpBackwardSelect` need to be different/configurable?  I think so because you want the select to be inclusive in both directions, which means `selectBackwards` would need to `beforeCharacter`.  So

```plaintext
jumpForward: beforeCharacter               // the defaults
selecForwardt: afterCharacter

jumpBackward: beforeCharacter
selecForwardt: beforeCharacter
```

**Comment by rainymatch** [comment]
1. Certainly! Works perfectly now for me with the suggested setting. Since my default setting is to jump before the character, it works correct without any additional setting for jumping and selecting backward. For forward I did as suggested: 
```
    {
        "key": "shift+alt+g",                 
        "command": "jump-and-select.jumpForwardSelect",
        "args": {
            "putCursorForward": "afterCharacter"
          }
    },
``` 
I don't see any need for any changes but relaxing the json schema, allowing for missing `text` argument. 

3. 

> I wonder if it would be possible to set a character (or something like alt+shift) that would trigger the reverse - which would be very handy? I'll investigate.

I see that in `multiMode`, pressing Tab will resolve in actually "writing" tab. If it was possible to make Tab change the direction of the multiMode, I think it would be great. Having information about the direction of the `multiMode` in the red notification in the bottom line, would be also helpful.

And you are totally right that I can simply enter the `multiMode` of the opposite direction while being in `multiMode` already. I guess visual feedback would be nice for that.
